src/semver.toit

```diff
@@ -1,195 +1,799 @@
+      -> SemanticVersion?:
+
+    parsed := ?
+    if peg:
+      parsed = (SemanticVersionPEGParser input
+        --accept-missing-minor=accept-missing-minor
+        --accept-missing-patch=accept-missing-patch
+        --accept-leading-zeros=accept-leading-zeros
+        --accept-v=accept-v).semantic-version
+        --consume-all
+    else:
+      parsed = (SemanticVersionTXTParser input
+        --accept-missing-minor=accept-missing-minor
+        --accept-missing-patch=accept-missing-patch
+        --accept-leading-zeros=accept-leading-zeros
+        --accept-v=accept-v).semantic-version
+        --consume-all
+    return parsed
+
+
+  // Construct object outright using switches.
+  constructor --.major/int --.minor/int=0 --.patch/int=0 --.pre-releases/List=[] --.build-metadata/List=[]:
+
+  // Construct object outright using ordered arguments.
+  constructor .major/any .minor/int=0 .patch/int=0 --.pre-releases/List=[] --.build-metadata/List=[]:
+
+  // Get version-core values in a list for simpler comparison.
+  triplet -> List: return [major, minor, patch]
+
+  /**
+  Creates a copy of the object with supplied properties changed.
+
+  Object is normally immutable, but easier with this helper which creates a
+    changed copy.
+  */
+  with --major /int? = null
+       --minor /int? = null
+       --patch /int? = null
+       --pre-release /List? = null
+       --build-metadata /List? = null:
+    return SemanticVersion
+        (major or this.major)
+        (minor or this.minor)
+        (patch or this.patch)
+        --pre-releases = (pre-releases or this.pre-releases)
+        --build-metadata = (build-metadata or this.build-metadata)
+
+  operator < other/SemanticVersion -> bool:
+    if compare-lists-less-than_ this.triplet other.triplet: return true
+    if compare-lists-less-than_ this.pre-releases other.pre-releases: return true
+    // Build-metadata should not be compared.
+    return false
+
+  operator > other/SemanticVersion -> bool:
+    return not this <= other
+
+  operator == other/SemanticVersion -> bool:
+    // Build-metadata should not be compared.
+    return (triplet == other.triplet) and (pre-releases == other.pre-releases)
+
+  operator >= other/SemanticVersion -> bool:
+    return not this < other
+
+  operator <= other/SemanticVersion -> bool:
+    return (this < other) or (this == other)
+
+  // Compare two lists using semver rules.  Works for version-core lists, as
+  // well as pre-release lists.
+  static compare-lists-less-than_ l1/List l2/List -> bool:
+    if (l2.size == 0) and (l1.size == 0): return false
+    if l2.size == 0: return true   // l1.size must be > 0 from earlier
+    if l1.size == 0: return false  // l2.size must be > 0 from earlier
+
+    l1.size.repeat:
+      // No matching cell in L2 to compare with L1.
+      if l2.size < (it + 1) : return false
+
+      // One string and one int:  Numeric always less than a string.
+      l1-numeric := (is-numeric_ l1[it])
```

Comment:
Don't extract the item from the list all the time.
Store it in a local variable.